### PR TITLE
Kp/develop

### PR DESCRIPTION
**Pull request by kprokofi**
Note:
* Additional scripts run_classification_* will be deleted
* If timm weights increase metrics, I merge it to common MobileNetv3 class

### CODE REVIEW

**Comment by kprokofi** [comment]
@sovrasov, The PR is ready for review

File: scripts/default_config.py [hunk 1]
```diff
@@ -136,16 +137,22 @@ def get_default_config():
...
     cfg.train.min_lr = 1e-5
     cfg.train.max_lr = 0.1
+    cfg.train.lr_decay_factor = 100
+    cfg.train.fixbase_epoch = 0
+    cfg.train.nbd = False
     cfg.train.patience = 5 # define how much epochs to wait for reduce on plateau
     cfg.train.multiplier = 10
     cfg.train.print_freq = 20  # print frequency
     cfg.train.seed = 5  # random seed
+    cfg.train.deterministic = False # degine to use cuda.deterministic
```
Comment (sovrasov) [inline scripts/default_config.py]: Typo in comment

File: scripts/script_utils.py [hunk 2]
```diff
@@ -146,10 +129,79 @@ def group_norm_symbolic(g, input_blob, num_groups, weight, bias, eps, cudnn_enab
...
 def is_config_parameter_set_from_command_line(cmd_line_opts, parameter_name):
     # Note that cmd_line_opts here should be compatible with
     # the function yacs.config.CfgNode.merge_from_list
     if not cmd_line_opts:
         return False
     key_names = cmd_line_opts[0::2]
     return (parameter_name in key_names)
+
+
+def put_on_the_device(model, use_gpu=True, gpu_num=1, num_aux_models=0, split_models=False):
```
Comment (sovrasov) [inline scripts/script_utils.py]: `put_main_model_on_the_device`

File: torchreid/utils/tools.py [hunk 3]
```diff
@@ -63,15 +63,19 @@ def read_yaml(fpath):
...
-    torch.backends.cudnn.benchmark = False
+    if deterministic:
+        torch.backends.cudnn.deterministic = True
+        torch.backends.cudnn.benchmark = False
     np.random.seed(seed)
     random.seed(seed)
     os.environ['PYTHONHASHSEED'] = str(seed)
 
+def worker_init_fn(worker_id):
+    np.random.seed(np.random.get_state()[1][0] + worker_id)
```
Comment (sovrasov) [inline torchreid/utils/tools.py:77]: I think here should be random.seed as well

File: torchreid/utils/lr_finder.py [hunk 4]
```diff
@@ -0,0 +1,134 @@
+from functools import partial
+
+import torch
+from torch_lr_finder import LRFinder
+import optuna
```
Comment (sovrasov) [inline torchreid/utils/lr_finder.py:5]: Was this package added to requirements.txt?

File: torchreid/utils/lr_finder.py [hunk 5]
```diff
@@ -0,0 +1,134 @@
...
+            lr_finder.reset()
+
+        return optim_lr
+
+    def optuna_optim(self):
+        study = optuna.create_study(study_name='classification task', direction="maximize", sampler=self.samplers[self.mode])
+        objective_partial = partial(self.engine.run, max_epoch=self.num_epochs, lr_finder=self.engine_cfg, start_eval=0, eval_freq=1,
+                                stop_callback=self.stop_callback)
+        try:
+            start_time = time.time()
```
Comment (sovrasov) [inline torchreid/utils/lr_finder.py:111]: What exception can occur here?

File: torchreid/models/mobilenetv3.py [hunk 6]
```diff
@@ -254,7 +306,7 @@ def init_pretrained_weights(model, key=''):
     def _get_torch_home():
         ENV_TORCH_HOME = 'TORCH_HOME'
         ENV_XDG_CACHE_HOME = 'XDG_CACHE_HOME'
-        DEFAULT_CACHE_DIR = '~/.cache'
+        DEFAULT_CACHE_DIR = './.cache'
```
Comment (sovrasov) [inline torchreid/models/mobilenetv3.py]: Why did you change the default cache dir? Now it would be harder to locate all the created cache folders, since their locations depend on current working dir.

File: torchreid/models/mobilenetv3.py [hunk 7]
```diff
@@ -240,7 +263,36 @@ def _initialize_weights(self):
...
+class MobileNetV3LargeTimm(MobileNetV3Base):
+    def __init__(self,
+                pretrained=False,
+                **kwargs):
+
+        super().__init__(**kwargs)
+        self.model = timm.create_model('mobilenetv3_large_100_miil_in21k', pretrained=pretrained, num_classes=self.num_classes)
+        self.use_angle_simple_linear = False
+        self.dropout = Dropout(**self.dropout_cls)
+        self.loss = 'softmax'
```
Comment (sovrasov) [inline torchreid/models/mobilenetv3.py]: What if `am_softmax` was initially passed to constructor? I think an error should be thrown in that case.

File: scripts/script_utils.py (hunk 2 above)
Comment (kprokofi) [reply scripts/script_utils.py]: why? In this function, the extra devices for aux models are initialized too

File: torchreid/utils/lr_finder.py (hunk 5 above)
Comment (kprokofi) [reply torchreid/utils/lr_finder.py:111]: if we want to terminate the lr search manually (cntrl+C) this would allow us to do that

Comment (kprokofi) [reply torchreid/utils/lr_finder.py:111]: and the training would proceed normally with the best learning rate chosen before terminating

File: torchreid/models/mobilenetv3.py (hunk 6 above)
Comment (kprokofi) [reply torchreid/models/mobilenetv3.py]: For some reason on the cluster, my home directory leaked through when downloading model weights... And I decided to change the directory

File: scripts/script_utils.py (hunk 2 above)
Comment (sovrasov) [reply scripts/script_utils.py]: It may additionally return indexes of devices for aux models if some particular non-default values of  kwargs are provided, but certainly puts the passed main model to some device.

File: torchreid/utils/tools.py (hunk 3 above)
Comment (kprokofi) [reply torchreid/utils/tools.py:77]: I guess, this problem concerns NumPy random generator only according to the article, but I added `random.seed` too

File: torchreid/models/mobilenetv3.py (hunk 6 above)
Comment (kprokofi) [reply torchreid/models/mobilenetv3.py]: Now, there is no problem anymore, but I agree that it would be harder to locate all the cache folders. Maybe just to write something like "/home/.pytorch/checkpoints"?

File: torchreid/utils/tools.py (hunk 3 above)
Comment (sovrasov) [reply torchreid/utils/tools.py:77]: The article just provides a sample implementation. Actually the problem concerns all not system-wide random engines since workers are standalone processes.

File: torchreid/models/mobilenetv3.py (hunk 6 above)
Comment (sovrasov) [reply torchreid/models/mobilenetv3.py]: On the cluster you can workaround that issue by exporting TORCH_HOME=valid directory for cache

Comment (kprokofi) [reply torchreid/models/mobilenetv3.py]: ok. I returned this back
